### DISCUSSION TITLE
Dynamically loading custom app with level 2 ncrc apps on INL HPC

### DISCUSSION THREAD

**Discussion by jfaulkner31** [idaholab/moose#25021] [Q&A Getting Started]
I have access to some level 2 ncrc applications (griffin, pronghorn, ect...) on INL HPC. I also have a custom application with my own kernels and FVBCs that I would like to use in my simulations. I have managed to get this running on my personal machine. The steps I take locally are: 

1. Install and compile ncrc application in conda channel. 
2. Compile MOOSE following the general instructions for compiling moose from source.
3. Compile my custom application.
4. Include something like the following in my input when running anyNcrcApp-opt:
```
[Problem]
    register_objects_from = 'FirebirdApp'
    library_path = '/home/jfaulkner31/projects/firebird/lib'
[]
```

But I am having a lot of difficulty doing the same on INL HPC to the point where I am wondering if what I want is even possible. I am  loading griffin/pronghorn as modules, compiling moose from the general INL-HPC instructions, compiling my own app using essentially the same makefile as I do locally, then hitting an error upon runtime that looks like this:
 
` symbol lookup error: ../../firebird/lib/libfirebird-opt.so.0: undefined symbol: _ZN16FunctorInterface14defaultFunctorIN11MetaPhysicL10DualNumberIdNS1_28SemiDynamicSparseNumberArrayIdmNS1_8NWrapperILm53EEEEELb1EEEEEPKN5Moose15FunctorEnvelopeIT_EERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE`

I imagine I might be able to do essentially the same as what I do locally through a conda environment, but that seems a bit silly considering there already are the proper modules available. I also imagine that I am missing a crucial step  when interacting with / loading modules. I am wondering if there is some kind of streamlined step by step process for what I want to accomplish or if I am indeed missing an important step. Any help is greatly appreciated! 

Cheers!

**Comment by GiudGiud** [comment]
Hello

It's awesome that this works on your machine.
On HPC, I would check every module to make sure the MOOSE version is the same between all the applications
In the header of a simulation you can see the commit number for moose

Guillaume

**Reply by jfaulkner31** [reply]
Hi Guillaume,

It looks like the two MOOSE versions are not the same: 

For griffin simulations I get:
```
Framework Information:
MOOSE Version:           git commit a24bf209dc on 2023-07-11
LibMesh Version:         dc964ef4e459ba7ce080f8d6c1f4cac939cfbfed
PETSc Version:           3.16.6
SLEPc Version:           3.16.2`
```
and for MOOSE / my custom app named Firebird:
```
MOOSE Version:           git commit d2ef20eecd on 2023-07-21
LibMesh Version:         dc964ef4e459ba7ce080f8d6c1f4cac939cfbfed
PETSc Version:           3.16.6
SLEPc Version:           3.16.2
```
However, I checked the same locally and it looks like each were built with with different framework versions back in May. Here is what appears in the headers in my local install that seems to work:

Firebird/MOOSE: 
```
Framework Information:
MOOSE Version:           git commit 0bd329c1be on 2023-06-05
LibMesh Version:         
PETSc Version:           3.16.6
SLEPc Version:           3.16.2
```
griffin-opt:
```
Framework Information:
MOOSE Version:           git commit 62c21c5da1 on 2023-05-25
LibMesh Version:         
PETSc Version:           3.16.6
SLEPc Version:           3.16.2
```
I have also noticed that when simply loading the griffin module before moose installation, I get all failing tests. But if I unload the griffin module before running tests, I get all passing tests in moose. OK, so I install moose without griffin loaded (but with the same mvapich2 version as griffin!) and then compile my app. I then run tests in my custom app and get all tests passing. I then load the griffin module, try to run tests again, and get all tests failing. 

I can also attempt to run the dynamically linked simulation as before using griffin-opt and get the same error:

```
channelTH0_externalLoop0: Parallelism:
channelTH0_externalLoop0:   Num Processors:        6
channelTH0_externalLoop0:   Num Threads:           1
channelTH0_externalLoop0: 

*** Warning, This code is deprecated and will be removed in future versions:
Deprecated Object: SolutionInitialCondition
This object will be removed on Mon Jul  1 01:00:00 2024
Replace SolutionInitialCondition with SolutionIC

griffin-opt: symbol lookup error: ../../firebird/lib/libfirebird-opt.so.0: undefined symbol: _ZN16FunctorInterface14defaultFunctorIN11MetaPhysicL10DualNumberIdNS1_28SemiDynamicSparseNumberArrayIdmNS1_8NWrapperILm53EEEEELb1EEEEEPKN5Moose15FunctorEnvelopeIT_EERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
```
I then completely comment out the FV interface kernel that is being inherited from the custom app - the simulation runs fine now, but without the interface kernel of course and with no complaints about the dynamically loaded app. 

Is there a way, when cloning moose, to use the same MOOSE version that was used in Griffin, so that both MOOSE versions are the same?

Thanks for the assistance!

**Reply by loganharbour** [reply]
You can see the version of moose you need:

> For griffin simulations I get:

```
Framework Information:
MOOSE Version: git commit a24bf20 on 2023-07-11
LibMesh Version: dc964ef4e459ba7ce080f8d6c1f4cac939cfbfed
PETSc Version: 3.16.6
SLEPc Version: 3.16.2
```

In the moose repository for your application, simply do:

```
git checkout a24bf209dced6a69d60604b746a61657ddda9e8d
```

which will put you in a detached (but valid) state of moose at the same commit as the other application.
